docs/check.go

```diff
@@ -0,0 +1,294 @@
+	}
+	return categoryNames, nil
+}
+
+var (
+	// general format of a relative link, where the link will be computed by a jekyll function encapsulated in {{ }}
+	relativeLinkRe = regexp.MustCompile(`\[.+\]\({{.*}}\)`)
+	// path encapsulated in ' ' could have an optional search query "?q=queryStr" and/or an optional anchor reference "#anchor"
+	relativeLinkPagePathRe = regexp.MustCompile(`\[.+\]\({{ '(?P<path>[\w-./]+)(\?q=\w+)?(#.+)?' | relativize_url }}\)`)
+)
+
+// checkFile parses the given markdown file and appends errors found in its contents
+func checkFile(mdFile string, ctx *checkContext) error {
+	f, err := os.Open(mdFile)
+	if err != nil {
+		return fmt.Errorf("error opening file at %v: %v", mdFile, err)
+	}
+	defer f.Close()
+
+	var headers []string
+	var relativeLinks []*relativeLink
+	inHeaderSection := true
+	scanner := bufio.NewScanner(f)
+	for i := 1; scanner.Scan(); i++ {
+		l := scanner.Text()
+		if inHeaderSection {
+			// first empty line ends the header section
+			if l == "" {
+				inHeaderSection = false
+			} else {
+				headers = append(headers, l)
+			}
+		}
+		if relativeLinkRe.MatchString(l) {
+			for _, lineMatches := range relativeLinkRe.FindAllStringSubmatch(l, -1) {
+				if len(lineMatches) != 1 {
+					return fmt.Errorf("expected to find exactly one string submatch but found %d in line %v", len(lineMatches), l)
+				}
+				relativeLinkStr := lineMatches[0]
+				if !relativeLinkPagePathRe.MatchString(relativeLinkStr) {
+					ctx.addError(mdFile, i, "relative link did not match expected pattern %q", relativeLinkStr)
+					continue
+				}
+				linkMatches := relativeLinkPagePathRe.FindStringSubmatch(relativeLinkStr)
+				if len(linkMatches) < 2 {
+					return fmt.Errorf("expected to find at least two string submatches but found %d = %v in link %v", len(linkMatches), linkMatches, relativeLinkStr)
+				}
+				// note that first is the full match, second is the named match
+				namedMatch := linkMatches[1]
+				if namedMatch == "" {
+					return fmt.Errorf("encountered empty named match when parsing link from %v", relativeLinkStr)
+				}
+				relativeLinks = append(relativeLinks, &relativeLink{
+					line: i,
+					path: namedMatch,
+				})
+			}
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("error scanning file: %v", err)
+	}
+
+	ctx.checkHeader(headers, mdFile)
+	ctx.addRelativeLinks(relativeLinks, mdFile)
+
+	return nil
+}
+
+const groupKey = "group"
+
+var headerKeys = StringSet{
+	"layout":   {},
+	"title":    {},
+	"nickname": {},
+	groupKey:   {},
+	"priority": {},
+}
+
+// checkHeader validates the header lines
+func (ctx *checkContext) checkHeader(headers []string, mdFile string) {
+	for i := 0; i < len(headers); i++ {
+		l := headers[i]
+		if i == 0 || i == len(headers)-1 {
+			if l != "---" {
+				ctx.addError(mdFile, i, "header section should be surrounded by --- but was %v", l)
+			}
+			continue
+		}
+		split := strings.Split(l, ": ")
```

Comment:
All content between `---` is yaml. We should use built-in deserializers if possible.

Reply:
same comment as above